src/cdomains/intDomain.ml

```diff
@@ -743,13 +778,36 @@ struct
     | `Bot -> `Bot
 
   let lift2 f x y = match x,y with
+    (* The good case: *)
+    | `Definite x, `Definite y ->
+      (try `Definite (f x y) with | Division_by_zero -> top ())
     (* We don't bother with exclusion sets: *)
-    | `Excluded _, _ -> top ()
-    | _, `Excluded _ -> top ()
+    | `Excluded (_, r), `Definite _
+    | `Definite _, `Excluded (_, r) ->
+      `Excluded (S.empty (), r)
+    | `Excluded (_, xr), `Excluded (_, yr) ->
+      check_identical_range xr yr;
+      `Excluded (S.empty (), xr)
+    | ` Bot, `Bot -> `Bot
+    | _ ->
+      (* If only one of them is bottom, we raise an exception that eval_rv will catch *)
+      raise (ArithmeticOnIntegerBot (Printf.sprintf "%s op %s" (short 80 x) (short 80 y)))
+
+  (* For the shift operations, CIL does not cast the right argument to the type of the left argument,    *)
+  (* so we should not warn about operations on different types here. The result has the type of the left *)
+  (* argument *)
+  let lift2_special f x y = match x,y with
     (* The good case: *)
     | `Definite x, `Definite y -> (try `Definite (f x y) with | Division_by_zero -> top ())
-    (* If any one of them is bottom, we return top *)
-    | _ -> top ()
+    (* We don't bother with exclusion sets: *)
+    | `Excluded (_, r), `Definite _
+    | `Definite _, `Excluded (_, r)
+    | `Excluded (_, r), `Excluded (_, _) ->
+      `Excluded (S.empty (), r)
```

Comment:
`Excluded (S.empty (), r)` was sound only because `r` was the full range of the left argument. When changing this, it is now unsound as well (on #95 as well, but added it here since #95 does not touch this piece of code).

```
x = 1;
if(top) {
   x =2;
}
// x→ Exc{0} [0,8]

y = x << 10; 
// y → Exc {} [0,8]
```

Reply:
Thanks. There might be several of those changes that need to be reverted.

Reply:
E.g. `lift2` as well